libs/feeds_processing/src/utils.rs

```diff
@@ -0,0 +1,612 @@
+                    warn!("Wrong value type reported by reporter {} for feed id {} slot {}! {} expected", kv.0, feed_id, slot, expected_feed_type.enum_type_to_string());
+                }
+            }
+            Err(_) => {
+                warn!(
+                    "Got error from reporter {} for feed id {} slot {}",
+                    kv.0, feed_id, slot
+                );
+            }
+        }
+    }
+    values
+}
+
+pub async fn perform_anomaly_detection(
+    feed_id: u32,
+    history: Arc<RwLock<FeedAggregateHistory>>,
+    candidate_value: f64,
+) -> Result<f64, anyhow::Error> {
+    let anomaly_detection_future = async move {
+        debug!("Get a read lock on history [feed {feed_id}]");
+        let history_lock = history.read().await;
+
+        // The first slice is from the current read position to the end of the array
+        // The second slice represents the segment from the start of the array up to the current write position if the buffer has wrapped around
+        let heap = history_lock
+            .get(feed_id)
+            .context("Missing key from History!")?;
+        let (first, last) = heap.as_slices();
+        let history_vec: Vec<&FeedType> =
+            first.iter().chain(last.iter()).map(|h| &h.value).collect();
+        let mut numerical_vec: Vec<f64> = history_vec
+            .iter()
+            .filter_map(|feed| {
+                if let FeedType::Numerical(value) = feed {
+                    Some(*value)
+                } else if let FeedType::Text(_) = feed {
+                    warn!("Anomaly Detection not implemented for FeedType::Text, skipping...");
+                    None
+                } else {
+                    warn!("Anomaly Detection does not recognize FeedType, skipping...");
+                    None
+                }
+            })
+            .collect();
+
+        drop(history_lock);
+        debug!("Release the read lock on history [feed {feed_id}]");
+
+        numerical_vec.push(candidate_value);
+
+        // Get AD prediction only if enough data is present
+        if numerical_vec.len() > AD_MIN_DATA_POINTS_THRESHOLD {
+            debug!("Starting anomaly detection for [feed {feed_id}]");
+            anomaly_detector_aggregate(numerical_vec).map_err(|e| anyhow!("{e}"))
+        } else {
+            Err(anyhow!(
+                "Skipping anomaly detection; numerical_vec.len() = {} threshold: {}",
+                numerical_vec.len(),
+                AD_MIN_DATA_POINTS_THRESHOLD
+            ))
+        }
+    };
+
+    tokio::task::Builder::new()
+        .name("anomaly_detection")
+        .spawn(anomaly_detection_future)
+        .context("Failed to spawn feed slots manager anomaly detection!")?
+        .await
+        .context("Failed to join feed slots manager anomaly detection!")?
+}
+
+type FeedId = u32;
+type FeedTimestamp = u64;
+type FeedIdToTimetamp = HashMap<FeedId, FeedTimestamp>;
+
+pub async fn validate_sigcheck(
```

Comment:
This function is not used anywhere.

Has it been tested in any way?

Same for validate.